File: Web/demo/MessageModal/network/sending_controller.py
import threading
import socket
import time

DEBUG = True


class SendingController(threading.Thread):
    def __init__(self, queue):
        super(SendingController, self).__init__(name='SendingController')
        self.sending_queue = queue
        self.thread_running = True

    def run(self):
        while self.thread_running:
            if not self.sending_queue.empty():
                target_url, packet = self.sending_queue.get()
                self.sent_to(target_url, packet)
            else:
                time.sleep(0.1)

    def stop(self):
        self.thread_running = False

    @staticmethod
    def sent_to(target_url, packet):
        s = socket.socket()
        s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        s.connect((target_url, 2333))
        s.sendall(packet.to_bytes())
        s.close()
        if DEBUG:
            print '[LOG][%s]: send a %s packet which uuid is %s to %s' % (
                time.ctime(), packet.get_command_string(), packet.uuid.encode('hex'), target_url)

File: MessageModel_c#/MessageModel/MessageModel/network/sending_controller.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

using System.Collections;
using System.Net;
using MessageModel.utils;
using System.Threading;
using System.Net.Sockets;
using MessageModel.packet;

namespace MessageModel.network
{
    class sending_controller
    {
        public static bool DEBUG = true;
        public Queue<Tuple<IPAddress, Packet>> sending_queue;
        bool threading_running = false;

        public sending_controller(Queue<Tuple<IPAddress, Packet>> queue)
        {
            this.sending_queue = queue;
            this.threading_running = true;
        }
        public void run()
        {
            while (this.threading_running)
            {
                if (sending_queue.Count != 0)
                {
                    Tuple<IPAddress, Packet> first = this.sending_queue.Dequeue();
                    IPAddress target_url = first.Item1;
                    Packet packet = first.Item2;
                    //Console.WriteLine(Encoding.UTF8.GetString(packet.Payload));
                    send_to(target_url, packet);
                }
                else
                    Thread.Sleep(100);
            }
        }

        public void stop()
        {
            this.threading_running = false;
        }

        public static void send_to(IPAddress target_url,Packet packet)
        {
            Socket s = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
            s.Connect(new IPEndPoint(target_url, 2333));
            s.Send(packet.to_bytes());
            s.Close();
            if (DEBUG)
                Console.WriteLine("[LOG][{0}]: send a {1} packet which uuid is {2} to {3}",DateTime.Now.ToString(),Encoding.UTF8.GetString(packet.get_command_string()),Encoding.UTF8.GetString(packet.Packet_uuid) ,target_url.ToString());
        }

    }
}
